### PR TITLE
Implement Exponential retry strategy

### PR DESCRIPTION
**Pull request by oneEyedSunday**
For https://github.com/frain-dev/convoy/issues/164

### CODE REVIEW

File: retrystrategies/retry.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+type RetryStrategy interface {
+	// NextDuration is how long we should wait before next retry
+	NextDuration(attempts uint64) time.Duration
+}
+
+func NewRetryStrategyFromMetadata(m datastore.Metadata) RetryStrategy {
+	if string(m.Strategy) == string(config.ExponentialBackoffStrategyProvider) {
+		// 0 to 5 seconds
+		return NewExponential([]uint{0, 10, 10, 100, 100, 500, 500, 3000, 3000, 5000})
```
Comment (oneEyedSunday) [inline retrystrategies/retry.go:18]: We probably should find a way to derive this from the IntervalSeconds property of the group

**Comment by oneEyedSunday** [comment]
I'm not quite sure how, but the bolt stuff is breaking the pipeline. FYI @Subomi

File: retrystrategies/retry_test.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+)
+
+func TestRetry_CreatesExponential(t *testing.T) {
+	m := datastore.Metadata{
+		Strategy:        "exponential-backoff",
+		RetryLimit:      20,
+		IntervalSeconds: 5,
+	}
+	var r RetryStrategy = NewRetryStrategyFromMetadata(m)
+	assert.NotEqual(t, r.NextDuration(1), r.NextDuration(3))
```
Comment (oneEyedSunday) [inline retrystrategies/retry_test.go:17]: I'm sure there's a way to confirm which implementation gets returned here. But for the life of me, I cant remember how that's done.

I remember it looking like:
`r.(ConcreteImplementation)`

**Comment by oneEyedSunday** [comment]
> --- FAIL: TestApplicationHandler_UpdateGroup (0.00s)
    --- FAIL: TestApplicationHandler_UpdateGroup/valid_group_update (0.00s)

Do i need to regenerate the mocks and push locally first?

As an aside, how do I run the mockgen locally?
```
ispoa@Sundays-MBP hookcamp % make mockgen 
go generate ./...
generate.go:3: running "mockgen": exec: "mockgen": executable file not found in $PATH
make: *** [mockgen] Error 1
```

**Comment by oneEyedSunday** [comment]
Can I get some help here, the issue here is because of the new field in the group config, we have a new result. 
Below, I've added a snippet of the golden Diff (DiffEngine.ClassicDiff). 

```
--- FAIL: TestApplicationHandler_UpdateGroup (0.00s)
    --- FAIL: TestApplicationHandler_UpdateGroup/valid_group_update (0.00s)
        application_test.go:29: Result did not match the golden fixture. Diff is below:
            
            --- Expected
            +++ Actual
            @@ -1,2 +1,2 @@
            -{"status":true,"message":"Group updated successfully","data":{"uid":"1234567890","name":"ABC_DEF_TEST_UPDATE","logo_url":"","config":{"strategy":{"type":"","default":{"intervalSeconds":0,"retryLimit":0}},"signature":{"header":"","hash":""},"disable_endpoint":false},"statistics":null}}
            +{"status":true,"message":"Group updated successfully","data":{"uid":"1234567890","name":"ABC_DEF_TEST_UPDATE","logo_url":"","config":{"strategy":{"type":"","default":{"intervalSeconds":0,"retryLimit":0},"exponentialBackoff":{"retryLimit":0}},"signature":{"header":"","hash":""},"disable_endpoint":false},"statistics":null}}
            
```

This is the new result of CreateGroup from the previous test
```

time="2022-01-19T03:14:19+01:00" level=warning msg="using noop realm for authentication: all requests will be authenticated with super_user role"
s {"status":true,"message":"Group updated successfully","data":{"uid":"1234567890","name":"ABC_DEF_TEST_UPDATE","logo_url":"","config":{"strategy":{"type":"","default":{"intervalSeconds":0,"retryLimit":0},"exponentialBackoff":{"retryLimit":0}},"signature":{"header":"","hash":""},"disable_endpoint":false},"statistics":null}}

```

For some reason, the mocks don't pick up the new field.

Another option is just making the new field a pointer, but we've not done that so far (see FileRealmOption and AuthConfiguration).

How do I fix this? I've run the mockgen locally, but it generates no changes.

cc @jirevwe @Subomi

File: retrystrategies/retry_test.go (hunk 2 above)
Comment (oneEyedSunday) [reply retrystrategies/retry_test.go]: resolved in https://github.com/frain-dev/convoy/pull/355/commits/5c2514044a001a8a20a53712ab0264da42df7bbe

**Comment by danvixent** [comment]
@oneEyedSunday do you still need help here?

**Comment by oneEyedSunday** [comment]
Yes, I thjnk the mocks are not picking up the new property (exponential
Backoff) so on update theres a diff.

**Comment by oneEyedSunday** [comment]
This is basically the failing test case

https://github.com/frain-dev/convoy/pull/355#issuecomment-1016018882

**Comment by oneEyedSunday** [comment]
so, thanks to @danvixent we figured we had to update the `.golden` files needed updating. I'd rebase then update the PR

**Review by jirevwe — APPROVED** [review]

**Review by danvixent — APPROVED** [review]

**Review by ogbanugot — APPROVED** [review]

**Pull request merged**